src/adapter/src/catalog.rs

```diff
@@ -215,19 +226,82 @@ impl Catalog {
     pub fn try_get_dataflow_metainfo(
         &self,
         id: &GlobalId,
-    ) -> Option<&DataflowMetainfo<OptimizerNotice>> {
+    ) -> Option<&DataflowMetainfo<Arc<OptimizerNotice>>> {
         self.plans.dataflow_metainfos.get(id)
     }
 
-    /// Drop all optimized and physical plans and `DataflowMetainfo`s for the item identified by
-    /// `id`.
+    /// Drop all optimized and physical plans and `DataflowMetainfo`s for the
+    /// item identified by `id`.
     ///
     /// Ignore requests for non-existing plans or `DataflowMetainfo`s.
+    ///
+    /// Return a set containing all dropped notices. Note that if for some
+    /// reason we end up with two identical notices being dropped by the same
+    /// call, the result will only contain only one instance of that notice.
     #[tracing::instrument(level = "trace", skip(self))]
-    pub fn drop_plans_and_metainfos(&mut self, id: GlobalId) {
-        self.plans.optimized_plan_by_id.remove(&id);
-        self.plans.physical_plan_by_id.remove(&id);
-        self.plans.dataflow_metainfos.remove(&id);
+    pub fn drop_plans_and_metainfos(
+        &mut self,
+        drop_ids: &BTreeSet<GlobalId>,
+    ) -> BTreeSet<Arc<OptimizerNotice>> {
+        // Collect dropped notices in this set.
+        let mut dropped_notices = BTreeSet::new();
+
+        // Remove plans and metainfo.optimizer_notices entries.
+        for id in drop_ids {
+            self.plans.optimized_plan_by_id.remove(id);
+            self.plans.physical_plan_by_id.remove(id);
+            if let Some(mut metainfo) = self.plans.dataflow_metainfos.remove(id) {
+                for n in metainfo.optimizer_notices.drain(..) {
+                    // Remove the corresponding notices_by_dep_id entries.
+                    for dep_id in n.dependencies.iter() {
+                        if let Some(notices) = self.plans.notices_by_dep_id.get_mut(dep_id) {
+                            notices.retain(|x| &n != x)
+                        }
+                    }
+                    dropped_notices.insert(n);
+                }
+            }
+        }
+
+        // Remove notices_by_dep_id entries.
+        for id in drop_ids {
+            if let Some(mut notices) = self.plans.notices_by_dep_id.remove(id) {
+                for n in notices.drain(..) {
+                    // Remove the corresponding metainfo.optimizer_notices entries.
+                    if let Some(item_id) = n.item_id.as_ref() {
+                        if let Some(metainfo) = self.plans.dataflow_metainfos.get_mut(item_id) {
+                            metainfo.optimizer_notices.retain(|x| &n != x)
```

Comment:
Uncovered

Reply:
We don't have notices that yield a dependency structure that is complex enough so I can produce an example to cover this line.